contract/test/test-crypto-xsnap.js

```diff
@@ -0,0 +1,177 @@
+// @ts-check
+import '@endo/init/debug.js';
+import { xsnap } from '@agoric/xsnap';
+import { encodeBase64 } from '@endo/base64';
+import test from 'ava';
+import proc from 'node:child_process';
+import fsMod from 'node:fs';
+import osMod from 'node:os';
+import tmpMod from 'tmp';
+import { pubkeyToAddress } from '../src/check-sig.js';
+
+const { freeze } = Object;
+
+// a TextDecoder has mutable state; only export the (pure) decode function
+/** @type { TextDecoder['decode'] } */
+export const decode = (decoder => decoder.decode.bind(decoder))(
+  new TextDecoder(),
+);
+
+/** @type { TextEncoder['encode'] } */
+export const encode = (encoder => encoder.encode.bind(encoder))(
+  new TextEncoder(),
+);
+
+/**
+ * @param {{
+ *   spawn?: typeof import('child_process').spawn,
+ *   os?: string,
+ *   fs?: import('fs'),
+ *   tmpName?: import('tmp')['tmpName'],
+ * }} [io]
+ * @returns {import('@agoric/xsnap').XSnapOptions & { messages: string[]}}
+ */
+export function options(io = {}) {
+  const {
+    spawn = proc.spawn,
+    os = osMod.type(),
+    fs = fsMod,
+    tmpName = tmpMod.tmpName,
+  } = io;
+
+  const messages = [];
+
+  /** @param {Uint8Array} message */
+  async function handleCommand(message) {
+    messages.push(decode(message));
+    return new Uint8Array();
+  }
+
+  return freeze({
+    name: 'xsnap test worker',
+    stderr: 'inherit',
+    stdout: 'inherit',
+    spawn,
+    fs: { ...fs, ...fs.promises, tmpName },
+    os,
+    handleCommand,
+    messages,
+  });
+}
+
+test('pubkeyToAddress matches cosmjs tests', t => {
+  const cases = [
+    {
+      prefix: 'cosmos',
+      pubkey: {
+        type: 'tendermint/PubKeySecp256k1',
+        value: 'AtQaCqFnshaZQp6rIkvAPyzThvCvXSDO+9AzbxVErqJP',
+      },
+      expected: 'cosmos1h806c7khnvmjlywdrkdgk2vrayy2mmvf9rxk2r',
+    },
+  ];
+  for (const { prefix, pubkey, expected } of cases) {
+    const actual = pubkeyToAddress(pubkey.value, prefix);
+    t.is(actual, expected);
+  }
+});
+
+test('Base64.encode', async t => {
+  const opts = options();
+  const vat = await xsnap(opts);
+  t.teardown(() => vat.terminate());
+  await vat.evaluate(`
+      const encoder = new TextEncoder();
+      globalThis.handleCommand = inputBuffer => {
+        const outputString = Base64.encode(inputBuffer);
+        const outputUint8Array = encoder.encode(outputString);
+        globalThis.issueCommand(outputUint8Array.buffer);
+      };
+    `);
+  const inputUint8Array = new TextEncoder().encode('Hello, World! 😃🌍');
+  const expectedOutputString = encodeBase64(inputUint8Array);
+  await vat.issueCommand(inputUint8Array);
+  t.deepEqual(opts.messages, [expectedOutputString]);
+});
+
+test('high resolution timer', async t => {
+  const opts = options();
+  const vat = await xsnap(opts);
+  t.teardown(() => vat.terminate());
+  await vat.evaluate(`
+    const send = it => issueCommand(new TextEncoder().encode(JSON.stringify(it)).buffer);
+
+    const t = performance.now();
```

Comment:
`performance.now()` is available using this sort of test where we ship the code over to XS using the `xsnap` API, but not when using `ava-xs`, which sets up an environment more like a vat, where no ambient authority is available.

Reply:
i think i read something about `performance.now()` potentially being harmful because high precision timestamps could be used in time-based attacks. i'll need to find a source though. 

i would think that it would be removed entirely in the event that it opens up an attack vector

Reply:
yes, `performance.now()` is a pointy instrument; you don't want to hand it out indiscriminately. But there's a time and a place for it. And that time/place is exactly benchmarks such as this one.